File: java-backend/src/main/java/com/forge/configurator/service/DrawingGenerationService.java
package com.forge.configurator.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Service
public class DrawingGenerationService {
    private static final Set<String> ALLOWED_BREAKERS = Set.of("ABB", "SCHNEIDER", "SIEMENS");

    private final String baseUrl;
    private final Duration timeout;
    private final HttpClient client;
    private final ObjectMapper objectMapper;

    public DrawingGenerationService(@Value("${app.drawing.solidworks-api-url:http://localhost:5100}") String baseUrl,
                                    @Value("${app.drawing.timeout-seconds:30}") long timeoutSeconds,
                                    ObjectMapper objectMapper) {
        this.baseUrl = baseUrl;
        this.timeout = Duration.ofSeconds(Math.max(1, timeoutSeconds));
        this.client = HttpClient.newBuilder().connectTimeout(this.timeout).build();
        this.objectMapper = objectMapper;
    }

    public DrawingResponse health() {
        return request("GET", "/api/health", null);
    }

    public DrawingResponse listJobs() {
        return request("GET", "/Drawings/jobs", null);
    }

    public DrawingResponse getJob(String jobId) {
        return request("GET", "/Drawings/jobs/" + encode(jobId), null);
    }

    public DrawingResponse listJobFiles(String jobId) {
        return request("GET", "/Drawings/jobs/" + encode(jobId) + "/files", null);
    }

    public DrawingResponse downloadJobFile(String jobId, String filename) {
        String path = "/Drawings/jobs/" + encode(jobId) + "/download?file=" + encode(filename);
        return request("GET", path, null);
    }

    public DrawingResponse createDrawing(String folderName, Integer panelCount, String circuitBreakerBrand) {
        if (folderName == null || folderName.isBlank() || folderName.length() > 200) {
            return DrawingResponse.error(422, "folderName must be 1..200 chars");
        }
        if (panelCount == null || panelCount < 1 || panelCount > 20) {
            return DrawingResponse.error(422, "panelCount must be an integer 1..20");
        }

        String brand = circuitBreakerBrand == null ? "" : circuitBreakerBrand.toUpperCase();
        if (!ALLOWED_BREAKERS.contains(brand)) {
            return DrawingResponse.error(422, "circuitBreakerBrand must be one of ABB, SCHNEIDER, SIEMENS");
        }

        return request("POST", "/Drawings/create", Map.of(
                "folderName", folderName,
                "panelCount", panelCount,
                "circuitBreakerBrand", brand
        ));
    }

    private DrawingResponse request(String method, String path, Map<String, Object> jsonBody) {
        try {
            String normalizedBase = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
            HttpRequest.Builder builder = HttpRequest.newBuilder(URI.create(normalizedBase + path))
                    .timeout(timeout)
                    .header("Accept", "application/json")
                    .header("ngrok-skip-browser-warning", "true");

            if (jsonBody != null) {
                String body = objectMapper.writeValueAsString(jsonBody);
                builder.header("Content-Type", "application/json");
                builder.method(method, HttpRequest.BodyPublishers.ofString(body));
            } else {
                builder.method(method, HttpRequest.BodyPublishers.noBody());
            }

            HttpResponse<byte[]> response = client.send(builder.build(), HttpResponse.BodyHandlers.ofByteArray());
            int status = response.statusCode();
            String contentType = response.headers().firstValue("content-type").orElse("application/octet-stream");
            byte[] body = response.body() == null ? new byte[0] : response.body();

            if (contentType.contains("application/json")) {
                Map<String, Object> json = body.length == 0
                    ? Map.of()
                    : readJsonMap(body);
                return new DrawingResponse(status < 400, status, false, null, json, null, contentType);
            }
            if (contentType.startsWith("text/")) {
                String text = new String(body, StandardCharsets.UTF_8);
                return new DrawingResponse(status < 400, status, false, null, text, null, contentType);
            }
            return new DrawingResponse(status < 400, status, false, null, null, body, contentType);
        } catch (Exception ex) {
            return new DrawingResponse(false, 503, true,
                    ex.getClass().getSimpleName().equals("HttpTimeoutException")
                            ? "SolidWorks API timed out"
                            : "SolidWorks API unreachable: " + ex.getMessage(),
                    null,
                    null,
                    "application/json");
        }
    }

    private String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> readJsonMap(byte[] body) throws java.io.IOException {
        return objectMapper.readValue(body, Map.class);
    }

    public record DrawingResponse(boolean ok,
                                  int status,
                                  boolean fallback,
                                  String error,
                                  Object body,
                                  byte[] bytes,
                                  String contentType) {
        public static DrawingResponse error(int status, String message) {
            return new DrawingResponse(false, status, false, message, null, null, "application/json");
        }
    }

    public List<String> allowedBreakers() {
        return List.copyOf(ALLOWED_BREAKERS);
    }
}

File: backend/src/services/configurator/drawingGenerationService.js
'use strict';

/**
 * drawingGenerationService.js — proxy to the SolidWorks drawing-generation API.
 *
 * Wraps the legacy `config/backend/app/routers/solidworks.py` proxy with:
 *   • Per-call timeout (default 30s, configurable via env)
 *   • Graceful 503 fallback when the upstream is unreachable
 *   • Automatic injection of `ngrok-skip-browser-warning: true` header
 *
 * NEVER throws on connection-refused / timeout — converts to
 * `{ ok: false, status: 503, fallback: true, error: '...' }` so the
 * controller can return a safe JSON response.
 *
 * NOTE: This is a thin pass-through. SLD diagrams are a SEPARATE
 * concept (see `ConfiguratorSldDocument`) and are NOT created here.
 */

const logger = require('../../utils/logger');

const SW_URL          = process.env.SOLIDWORKS_API_URL || 'http://localhost:5100';
const SW_TIMEOUT_MS   = Number(process.env.SOLIDWORKS_TIMEOUT_MS) || 30000;

const DEFAULT_HEADERS = {
  Accept: 'application/json',
  'ngrok-skip-browser-warning': 'true',
};

async function _request(method, pathSuffix, { json, query } = {}) {
  let url = `${SW_URL.replace(/\/$/, '')}${pathSuffix}`;
  if (query && Object.keys(query).length) {
    const qs = new URLSearchParams(query).toString();
    url += (url.includes('?') ? '&' : '?') + qs;
  }
  const ac = new AbortController();
  const timer = setTimeout(() => ac.abort(), SW_TIMEOUT_MS);
  try {
    const res = await fetch(url, {
      method,
      headers: {
        ...DEFAULT_HEADERS,
        ...(json ? { 'Content-Type': 'application/json' } : {}),
      },
      body: json ? JSON.stringify(json) : undefined,
      signal: ac.signal,
    });
    const contentType = res.headers.get('content-type') || '';
    let body = null;
    if (contentType.includes('application/json')) {
      body = await res.json().catch(() => null);
    } else if (contentType.startsWith('text/')) {
      body = await res.text().catch(() => null);
    } else {
      body = await res.arrayBuffer().catch(() => null);
    }
    return { ok: res.ok, status: res.status, body, headers: res.headers, contentType };
  } catch (err) {
    const isAbort = err.name === 'AbortError';
    logger.warn({ err: err.message, url, isAbort },
      '[configurator/drawing] upstream unreachable — returning fallback');
    return {
      ok: false,
      status: 503,
      fallback: true,
      error: isAbort ? 'SolidWorks API timed out' : `SolidWorks API unreachable: ${err.message}`,
    };
  } finally {
    clearTimeout(timer);
  }
}

async function health() {
  return _request('GET', '/api/health');
}

async function listJobs() {
  return _request('GET', '/Drawings/jobs');
}

async function getJob(jobId) {
  return _request('GET', `/Drawings/jobs/${encodeURIComponent(jobId)}`);
}

async function listJobFiles(jobId) {
  return _request('GET', `/Drawings/jobs/${encodeURIComponent(jobId)}/files`);
}

async function downloadJobFile(jobId, filename) {
  return _request('GET', `/Drawings/jobs/${encodeURIComponent(jobId)}/download`, {
    query: { file: filename },
  });
}

const ALLOWED_BREAKERS = ['ABB', 'SCHNEIDER', 'SIEMENS'];

async function createDrawing({ folderName, panelCount, circuitBreakerBrand }) {
  if (!folderName || folderName.length < 1 || folderName.length > 200) {
    const err = new Error('folderName must be 1..200 chars'); err.status = 422; throw err;
  }
  if (!Number.isInteger(panelCount) || panelCount < 1 || panelCount > 20) {
    const err = new Error('panelCount must be an integer 1..20'); err.status = 422; throw err;
  }
  const brand = String(circuitBreakerBrand || '').toUpperCase();
  if (!ALLOWED_BREAKERS.includes(brand)) {
    const err = new Error(`circuitBreakerBrand must be one of ${ALLOWED_BREAKERS.join(', ')}`);
    err.status = 422; throw err;
  }

  return _request('POST', '/Drawings/create', {
    json: { folderName, panelCount, circuitBreakerBrand: brand },
  });
}

module.exports = {
  health,
  listJobs,
  getJob,
  listJobFiles,
  downloadJobFile,
  createDrawing,
  ALLOWED_BREAKERS,
};
